### PR TITLE
Value encoding

### PR DESCRIPTION
**Pull request by redxaxder**
Different serializations for Value
    
The Value type is used in both TxOuts and in the Mint field of the transaction body, but negative values are only allowed in the latter. The rules already enforce this, but now the serialization does too.
    
Resolves: CAD-1667 and CAD-2148

### CODE REVIEW

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 1]
```diff
@@ -140,24 +147,77 @@ instance Era era => Val (Value era) where
...
+adaAssetName = AssetID mempty
+
+adaPolicyId :: forall era. Era era => PolicyID era
+adaPolicyId =  PolicyID . ScriptHash . Hash.UnsafeHash $ SBS.pack bytes
+  where
+  hashlen = fromIntegral $
+     Hash.sizeHash (Proxy :: Proxy (Crypto.ADDRHASH (Crypto era)))
+  ada = [0x0A,0xDA]
+  bytes = if hashlen <= 2
+    then take 2 ada
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: Is hashlen is 1, then we take 2?

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 2]
```diff
@@ -140,24 +147,77 @@ instance Era era => Val (Value era) where
...
+  case tt of
+    TypeUInt -> decodeIntegerValue
+    TypeNInt -> decodeIntegerValue -- TODO: Separate nongegative and allowed-negative
+    TypeMapLen -> decodeMapValue
+    TypeMapLen64 -> decodeMapValue
+    TypeMapLenIndef -> decodeMapValue
+    _ -> cborError $ DecoderErrorCustom "Value" "expected map or uint"
+
+
+integerToWord64 :: Integer -> Word64
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: Unused?

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 3]
```diff
@@ -204,42 +264,28 @@ insert ::
...
-            then link2 l1 l2
-            else link pid (Map.singleton aid new) l1 l2
-        )
+insert combine pid aid new (Value cn m1) = Value cn $ 
+  insert' combine pid aid new m1
 
 -- ========================================================
 
+-- | Remove 0 assets from a map
+prune :: Map (PolicyID era) (Map AssetID Integer)
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: I guess this is perhaps a bit slower than the old method?

**Review by nc6 — COMMENTED** [review]
Just left comments since this is still marked as draft.

**Review by JaredCorduan — APPROVED** [review]
LGTM!

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 4]
```diff
@@ -19,18 +21,18 @@ module Cardano.Ledger.Mary.Value
...
 where
 
-import Cardano.Binary
-  ( FromCBOR,
-    ToCBOR,
-    encodeListLen,
-    fromCBOR,
-    toCBOR,
-  )
+import Cardano.Binary (Decoder, Encoding, FromCBOR, ToCBOR, TokenType (..), decodeInt64, decodeWord64, fromCBOR, peekTokenType, toCBOR)
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: Can we avoid such very long lines?

File: shelley/chain-and-ledger/shelley-spec-ledger-test/.ghcid [hunk 5]
```diff
@@ -1 +1 @@
--c "nix-shell ../../../shell.nix --run \"cabal repl -f development shelley-spec-ledger-test\"" -o ghcid.txt
+-c "nix-shell ../../../shell.nix --run \"cabal repl -f development test:shelley-spec-ledger-test\"" -o ghcid.txt
```
Comment (nc6) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/.ghcid]: Is this change intentional?

**Review by nc6 — APPROVED** [review]
Looks pretty good to me, couple of minor comments only

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs [hunk 6]
```diff
@@ -362,4 +362,7 @@ translateValue :: Era era => Coin -> Value era
 translateValue = Val.inject
 
 translateCompactValue :: Era era => CompactForm Coin -> CompactForm (Value era)
-translateCompactValue = toCompact . translateValue . fromCompact
+translateCompactValue = assume . toCompact . translateValue . fromCompact
+  where
+    assume Nothing = error "impossible error: compact coin is out of range"
+    assume (Just x) = x
```
Comment (mrBliss) [inline shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: Suggestion:
```suggestion
translateCompactValue =
    fromMaybe (error msg) . toCompact . translateValue . fromCompact
  where
    msg = "impossible error: compact coin is out of range"
```

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs (hunk 4 above)
Comment (mrBliss) [reply shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: Doesn't ormolu do something about this?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxBody.hs [hunk 7]
```diff
@@ -483,7 +483,9 @@ pattern TxOut addr vl <-
   (viewCompactTxOut -> (addr, vl))
   where
     TxOut addr vl =
-      TxOutCompact (compactAddr addr) (toCompact vl)
+      let assume Nothing = error $ "illegal value in txout: " <> show vl
+          assume (Just x) = x
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxBody.hs]: Similar suggestion about `fromMaybe`

**Review by mrBliss — COMMENTED** [review]
Could you fix the PR description before merging? It would be very confusing if the current message ends up in the history. You could, for example, copy the commit message (also rename the PR to match the commit's title).

File: shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs (hunk 4 above)
Comment (nc6) [reply shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: Ormolu basically has two choices: one line, or split them. If you add a random line break somewhere in there, ormolu will then format the whole list. But it allows you to choose not to add any line breaks, as in this case

Comment (JaredCorduan) [reply shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: I'll break it up

File: shelley/chain-and-ledger/shelley-spec-ledger-test/.ghcid (hunk 5 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/shelley-spec-ledger-test/.ghcid]: probably not, I'll remove it
